src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -265,22 +262,82 @@ static ImmutableArray<string> AttachComplexTypes(ImmutableArray<string> receiver
             }
         }
 
-        private static void GetExtensionMethodItemsForSymbolsFromDifferentCompilation(
-            int position,
-            SemanticModel semanticModel,
-            ITypeSymbol receiverTypeSymbol,
+        private static async Task<ImmutableArray<IMethodSymbol>?> GetExtensionMethodSymbolsFromProjectAsync(
+            Project project,
+            bool isCurrentProject,
+            bool forceIndexCreation,
+            SymbolComputationParameters parameters,
+            CancellationToken cancellationToken)
+        {
+            var cacheEntry = await GetCacheEntryAsync(project, loadOnly: !forceIndexCreation, parameters.CacheService, cancellationToken).ConfigureAwait(false);
+            if (!cacheEntry.HasValue)
+            {
+                // Returns null to indicate index not ready
+                return null;
+            }
+
+            if (!cacheEntry.Value.ContainsExtensionMethod)
+            {
+                return ImmutableArray<IMethodSymbol>.Empty;
+            }
+
+            var currentAssembly = parameters.SemanticModel.Compilation.Assembly;
+            var filter = CreateAggregatedFilter(parameters.ReceiverTypeNames, cacheEntry.Value);
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var assembly = compilation.Assembly;
+            var internalsVisible = currentAssembly.IsSameAssemblyOrHasFriendAccessTo(assembly);
+
+            var matchingMethodSymbols = GetPotentialMatchingSymbolsFromAssembly(
+                compilation.Assembly, filter, parameters.NamespaceInScope, internalsVisible, cancellationToken);
+
+            return isCurrentProject
+                ? GetExtensionMethodsForSymbolsFromSameCompilation(
+                    matchingMethodSymbols, parameters, cancellationToken)
+                : GetExtensionMethodsForSymbolsFromDifferentCompilation(
+                    matchingMethodSymbols, parameters, cancellationToken);
+        }
+
+        private static async Task<ImmutableArray<IMethodSymbol>?> GetExtensionMethodSymbolsFromPeReferenceAsync(
+            PortableExecutableReference peReference,
+            bool forceIndexCreation,
+            SymbolComputationParameters parameters,
+            CancellationToken cancellationToken)
+        {
+            var index = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(parameters.Solution, peReference, loadOnly: !forceIndexCreation, cancellationToken).ConfigureAwait(false);
+            if (index == null)
+            {
+                // Returns null to indicate index not ready
+                return null;
+            }
+
+            if (index.ContainsExtensionMethod && parameters.SemanticModel.Compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assembly)
+            {
+                var filter = CreateAggregatedFilter(parameters.ReceiverTypeNames, index);
```

Comment:
the terms "filter" or "aggregated filter" aresn't really conveying to me what thsi is for.

Reply:
I have added some comments